### PR TITLE
perf: preprocess tc instances to avoid nested tc

### PR DESCRIPTION
**Pull request by gebner**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by gebner** [comment]
!bench

File: Mathlib/Util/Commutes.lean [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import Lean
+import Std.Data.Option.Basic
+import Std.Tactic.Lint
+
+/-!
+
+# Auxiliary type class to avoid nested TC problems
+
+This file defines a `[@Commutes C inst₁ inst₂]` type class meaning that the
+instances `inst₁ : C` and `inst₂ : C` are defeq.  This allows us to write instances like:
```
Comment (eric-wieser) [inline Mathlib/Util/Commutes.lean:15]: Presumably this is enforced socially and not by lean itself; it looks to me like the typeclass only enforces that they're propositionally equal?

Comment (gebner) [reply Mathlib/Util/Commutes.lean:15]: The only instance for `Commutes` witnesses definitional equality.  It's not meant to be extended.

Comment (eric-wieser) [reply Mathlib/Util/Commutes.lean:15]: Oh wow that's a neat trick! I totally missed that `Commutes.mk` was an instance.

**Comment by gebner** [comment]
I view this PR as a slightly experimental hack, both to see how well the approach works, as well as something that we could maybe ship today (and most likely allow us to enable eta).  The same issue arises with simp, and in aesop, and really anything that tries to apply lemmas in any form.  It would be great if we could fix this directly in unification, but even if had a perfect solution today, it would take a couple of weeks to get it in.

File: Mathlib/Util/Commutes.lean [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        let t ← inferType e
+        let u ← getLevel t
+        return mkApp3 (.const ``Commutes [u]) t y e)) fun cs => do
+  mkForallFVars (xs ++ ys ++ cs) ty
+
+open private declValToTerm from Lean.Elab.MutualDef in
+open Lean Elab Parser Command in
+/--
+Similar to `instance`, but generalizes nonatomic instance terms in the
+conclusion.  Only use for `Prop`-valued classes.
```
Comment (eric-wieser) [inline Mathlib/Util/Commutes.lean:91]: Can you enforce this with an error message?

**Comment by gebner** [comment]
!bench

**Comment by gebner** [comment]
That is a very specific improvement!

Unfortunately it is not enough to enable eta, as it doesn't solve `Semiring { x // 0 ≤ x }`, which still times out.

**Comment by YaelDillies** [comment]
I am closing this PR since it seems it has fulfilled its purpose. Feel free to reopen it.

**Pull request closed without merging**